File: src/harness/providers/codex/app-server.ts
import { readFile } from "node:fs/promises";

import type { HarnessResult } from "../../events.js";
import type { AgentRunSpec, HarnessRunHooks } from "../../types.js";
import { spawnManagedChildProcess } from "../../../process/managed-child.js";
import {
  buildCodexAppServerRequest,
  codexClientVersion,
  combineCodexPrompt,
} from "./request.js";
import { mapCodexAppServerNotification } from "./app-notifications.js";
import { classifyCodexFailure } from "./failures.js";
import {
  asRecord,
  stringField,
} from "./fields.js";
import { toHarnessResult } from "./result.js";
import type { CodexRunState } from "./types.js";

interface JsonRpcResponse {
  id: number | string;
  result?: unknown;
  error?: {
    message?: string;
    code?: number;
    data?: unknown;
  };
}

export interface JsonRpcNotification {
  method: string;
  params?: unknown;
}

interface PendingRequest {
  resolve: (value: unknown) => void;
  reject: (err: Error) => void;
}

const CODEX_APP_SERVER_RPC_TIMEOUT_MS = 30_000;
const CODEX_APP_SERVER_RPC_TIMEOUT_ENV =
  "CODEALMANAC_CODEX_APP_SERVER_RPC_TIMEOUT_MS";
const CODEX_APP_SERVER_TURN_TIMEOUT_MS = 30 * 60_000;
const CODEX_APP_SERVER_TURN_TIMEOUT_ENV =
  "CODEALMANAC_CODEX_APP_SERVER_TURN_TIMEOUT_MS";

export async function runCodexAppServer(
  spec: AgentRunSpec,
  hooks?: HarnessRunHooks,
): Promise<HarnessResult> {
  const request = buildCodexAppServerRequest(spec);
  const rpcTimeoutMs = codexAppServerRpcTimeoutMs(request.env);
  const turnTimeoutMs = codexAppServerTurnTimeoutMs(request.env);
  return new Promise((resolve) => {
    const managed = spawnManagedChildProcess(request.command, request.args, {
      cwd: request.cwd,
      env: request.env,
      stdio: ["pipe", "pipe", "pipe"],
    });
    const child = managed.child;
    const pending = new Map<string, PendingRequest>();
    const state: CodexRunState = { success: false, result: "" };
    const eventWrites: Promise<void>[] = [];
    let nextRequestId = 1;
    let stdoutBuf = "";
    let stderr = "";
    let settled = false;
    let activeTurnId: string | undefined;
    let turnTimeout: NodeJS.Timeout | undefined;
    const removeSignalHandlers = installSignalHandlers((signal) => {
      fail(`Codex app-server interrupted by ${signal}`);
    });

    const finish = async (result: HarnessResult): Promise<void> => {
      if (settled) return;
      settled = true;
      removeSignalHandlers();
      if (turnTimeout !== undefined) {
        clearTimeout(turnTimeout);
        turnTimeout = undefined;
      }
      for (const entry of pending.values()) {
        entry.reject(new Error("Codex app-server run finished"));
      }
      pending.clear();
      await Promise.allSettled(eventWrites);
      const cleanupError = await terminateManagedChildSafely(managed);
      if (cleanupError !== undefined) {
        await hooks?.onEvent?.({
          type: "error",
          error: cleanupError,
        });
      }
      resolve(result);
    };

    const fail = (raw: string): void => {
      if (settled) return;
      const failure = classifyCodexFailure(raw);
      state.success = false;
      state.error = failure.message;
      state.failure = failure;
      eventWrites.push(
        hooks?.onEvent?.({ type: "error", error: failure.message, failure }) ??
          Promise.resolve(),
      );
      void finish(toHarnessResult(state));
    };

    const startTurnWatchdog = (): void => {
      if (settled) return;
      if (turnTimeout !== undefined) clearTimeout(turnTimeout);
      turnTimeout = setTimeout(() => {
        fail(`Codex app-server turn timed out after ${turnTimeoutMs}ms`);
      }, turnTimeoutMs);
    };

    const write = (message: Record<string, unknown>): void => {
      child.stdin?.write(`${JSON.stringify(message)}\n`);
    };

    const requestRpc = (method: string, params?: unknown): Promise<unknown> => {
      const id = nextRequestId++;
      return new Promise((requestResolve, requestReject) => {
        const timeout = setTimeout(() => {
          pending.delete(String(id));
          requestReject(
            new Error(
              `Codex app-server ${method} timed out after ${rpcTimeoutMs}ms`,
            ),
          );
        }, rpcTimeoutMs);
        pending.set(String(id), {
          resolve: (value) => {
            clearTimeout(timeout);
            requestResolve(value);
          },
          reject: (err) => {
            clearTimeout(timeout);
            requestReject(err);
          },
        });
        write({
          id,
          method,
          ...(params !== undefined ? { params } : {}),
        });
      });
    };

    const respond = (id: string | number, result: unknown): void => {
      write({
        id,
        result,
      });
    };

    const respondError = (id: string | number, code: number, message: string): void => {
      write({
        id,
        error: {
          code,
          message,
        },
      });
    };

    const respondUnsupported = (id: string | number, method: string): void => {
      respondError(
        id,
        -32601,
        `Almanac does not handle Codex app-server request ${method}`,
      );
    };

    const handleResponse = (message: JsonRpcResponse): void => {
      const item = pending.get(String(message.id));
      if (item === undefined) return;
      pending.delete(String(message.id));
      if (message.error !== undefined) {
        item.reject(new Error(message.error.message ?? "Codex app-server request failed"));
        return;
      }
      item.resolve(message.result);
    };

    const handleNotification = (message: JsonRpcNotification): void => {
      const turnId = stringField(asRecord(message.params), "turnId");
      if (
        activeTurnId === undefined &&
        turnId !== undefined &&
        isRootThreadNotification(message, state)
      ) {
        activeTurnId = turnId;
      }
      const isRootCompletion = isRootTurnCompletion(message, state, activeTurnId);
      const events = mapCodexAppServerNotification(message, state, {
        activeTurnId,
        rootThreadId: state.rootThreadId,
        rootTurnId: state.rootTurnId,
        isRootCompletion,
      });
      for (const event of events) {
        eventWrites.push(hooks?.onEvent?.(event) ?? Promise.resolve());
      }
      if (message.method === "turn/completed") {
        if (isRootCompletion === false) {
          return;
        }
        state.success = state.failure === undefined;
        state.turns = 1;
        eventWrites.push(
          hooks?.onEvent?.({
            type: "done",
            result: state.result,
            providerSessionId: state.providerSessionId,
            turns: state.turns,
            usage: state.usage,
            error: state.error,
            failure: state.failure,
            sourceThreadId: state.resultSourceThreadId,
            sourceTurnId: state.resultSourceTurnId,
            sourceRole: state.resultSourceRole,
          }) ?? Promise.resolve(),
        );
        void finish(toHarnessResult(state));
      }
    };

    const handleMessage = (message: unknown): void => {
      if (message === null || typeof message !== "object") return;
      const record = message as Record<string, unknown>;
      if ("id" in record && "method" in record) {
        respondToServerRequest(
          record.id as string | number,
          String(record.method),
          respond,
          respondError,
          respondUnsupported,
        );
        return;
      }
      if ("id" in record) {
        handleResponse(record as unknown as JsonRpcResponse);
        return;
      }
      if ("method" in record) {
        handleNotification(record as unknown as JsonRpcNotification);
      }
    };

    const flushLines = (): void => {
      let idx = stdoutBuf.indexOf("\n");
      while (idx !== -1) {
        const rawLine = stdoutBuf.slice(0, idx);
        stdoutBuf = stdoutBuf.slice(idx + 1);
        const line = rawLine.trim();
        if (line.length > 0) {
          try {
            handleMessage(JSON.parse(line) as unknown);
          } catch {
            // Ignore non-JSON chatter; stderr is captured for failures.
          }
        }
        idx = stdoutBuf.indexOf("\n");
      }
    };

    child.stdout?.on("data", (chunk) => {
      stdoutBuf += chunk.toString("utf8");
      flushLines();
    });
    child.stderr?.on("data", (chunk) => {
      stderr += chunk.toString("utf8");
    });
    child.on("error", (err: NodeJS.ErrnoException) => {
      fail(err.code === "ENOENT" ? `${request.command} not found on PATH` : err.message);
    });
    child.on("close", (code) => {
      if (settled) return;
      flushLines();
      const firstStderr = stderr.trim().split("\n")[0];
      fail(
        firstStderr !== undefined && firstStderr.length > 0
          ? firstStderr
          : `${request.command} app-server exited ${code ?? 1}`,
      );
    });

    void (async () => {
      try {
        await requestRpc("initialize", {
          clientInfo: {
            name: "codealmanac",
            title: "Almanac",
            version: codexClientVersion(),
          },
          capabilities: {
            experimentalApi: true,
          },
        });
        const thread = asRecord(
          await requestRpc("thread/start", {
            cwd: spec.cwd,
            model: spec.provider.model ?? null,
            approvalPolicy: "never",
            sandbox: "workspace-write",
            developerInstructions: spec.systemPrompt ?? null,
            ephemeral: spec.providerSession?.persistence === "ephemeral",
          }),
        );
        const threadObj = asRecord(thread.thread);
        const threadId = stringField(threadObj, "id");
        if (threadId === undefined) {
          throw new Error("Codex app-server thread/start did not return a thread id");
        }
        state.providerSessionId = threadId;
        state.rootThreadId = threadId;
        eventWrites.push(
          hooks?.onEvent?.({ type: "provider_session", providerSessionId: threadId }) ??
            Promise.resolve(),
        );
        const outputSchema = await readOutputSchema(spec.output?.schemaPath);
        const turn = asRecord(
          await requestRpc("turn/start", {
            threadId,
            cwd: spec.cwd,
            input: [
              {
                type: "text",
                text: combineCodexPrompt({ ...spec, systemPrompt: undefined }),
                text_elements: [],
              },
            ],
            approvalPolicy: "never",
            sandboxPolicy: {
              type: "workspaceWrite",
              writableRoots: [spec.cwd],
              networkAccess: connectorNetworkAccess(spec),
              excludeTmpdirEnvVar: false,
              excludeSlashTmp: false,
            },
            model: spec.provider.model ?? null,
            effort: spec.provider.effort ?? null,
            outputSchema,
          }),
        );
        activeTurnId = stringField(asRecord(turn.turn), "id");
        state.rootTurnId = activeTurnId;
        startTurnWatchdog();
      } catch (err: unknown) {
        fail(err instanceof Error ? err.message : String(err));
      }
    })();
  });
}

function connectorNetworkAccess(spec: AgentRunSpec): boolean {
  return (spec.connectors?.length ?? 0) > 0;
}

function installSignalHandlers(onSignal: (signal: NodeJS.Signals) => void): () => void {
  const signals: NodeJS.Signals[] = ["SIGINT", "SIGTERM", "SIGHUP"];
  const handlers = signals.map((signal) => {
    const handler = () => onSignal(signal);
    process.once(signal, handler);
    return { signal, handler };
  });
  return () => {
    for (const { signal, handler } of handlers) {
      process.off(signal, handler);
    }
  };
}

async function terminateManagedChildSafely(
  managed: ReturnType<typeof spawnManagedChildProcess>,
): Promise<string | undefined> {
  try {
    await managed.terminate();
    return undefined;
  } catch (err: unknown) {
    return `Provider process cleanup failed: ${err instanceof Error ? err.message : String(err)}`;
  }
}

function isRootTurnCompletion(
  message: JsonRpcNotification,
  state: CodexRunState,
  activeTurnId: string | undefined,
): boolean | undefined {
  if (message.method !== "turn/completed") return undefined;
  const params = asRecord(message.params);
  const completedTurnId = stringField(params, "turnId");
  const completedThreadId = stringField(params, "threadId");
  const isRootTurn =
    (state.rootTurnId !== undefined && completedTurnId === state.rootTurnId) ||
    (state.rootTurnId === undefined &&
      activeTurnId !== undefined &&
      completedTurnId === activeTurnId);
  const isRootThread =
    state.rootThreadId !== undefined && completedThreadId === state.rootThreadId;
  return isRootTurn || isRootThread;
}

function isRootThreadNotification(
  message: JsonRpcNotification,
  state: CodexRunState,
): boolean {
  const completedThreadId = stringField(asRecord(message.params), "threadId");
  return (
    state.rootThreadId !== undefined &&
    completedThreadId === state.rootThreadId
  );
}

function respondToServerRequest(
  id: string | number,
  method: string,
  respond: (id: string | number, result: unknown) => void,
  respondError: (id: string | number, code: number, message: string) => void,
  respondUnsupported: (id: string | number, method: string) => void,
): void {
  switch (method) {
    case "item/commandExecution/requestApproval":
      respond(id, { decision: "decline" });
      return;
    case "item/fileChange/requestApproval":
      respond(id, { decision: "decline" });
      return;
    case "execCommandApproval":
    case "applyPatchApproval":
      respond(id, { decision: "denied" });
      return;
    case "item/tool/requestUserInput":
      respond(id, { answers: {} });
      return;
    case "mcpServer/elicitation/request":
      respond(id, { action: "decline", content: null, _meta: null });
      return;
    case "item/tool/call":
      respond(id, { contentItems: [], success: false });
      return;
    case "item/permissions/requestApproval":
      respond(id, {
        permissions: {},
        scope: "turn",
        strictAutoReview: true,
      });
      return;
    case "account/chatgptAuthTokens/refresh":
      respondError(
        id,
        -32001,
        "Almanac does not manage ChatGPT auth tokens for Codex app-server.",
      );
      return;
    default:
      respondUnsupported(id, method);
  }
}

async function readOutputSchema(schemaPath: string | undefined): Promise<unknown> {
  if (schemaPath === undefined) return null;
  const raw = await readFile(schemaPath, "utf8");
  return JSON.parse(raw) as unknown;
}

function codexAppServerRpcTimeoutMs(env: NodeJS.ProcessEnv): number {
  return parsePositiveEnvInt(
    env[CODEX_APP_SERVER_RPC_TIMEOUT_ENV],
    CODEX_APP_SERVER_RPC_TIMEOUT_MS,
  );
}

function codexAppServerTurnTimeoutMs(env: NodeJS.ProcessEnv): number {
  return parsePositiveEnvInt(
    env[CODEX_APP_SERVER_TURN_TIMEOUT_ENV],
    CODEX_APP_SERVER_TURN_TIMEOUT_MS,
  );
}

function parsePositiveEnvInt(value: string | undefined, fallback: number): number {
  if (value === undefined) return fallback;
  const parsed = Number(value);
  if (!Number.isFinite(parsed) || parsed <= 0) return fallback;
  return parsed;
}

File: src/codealmanac/integrations/harnesses/codex/app_server.py
import time

from pydantic import JsonValue

from codealmanac import __version__
from codealmanac.integrations.harnesses.codex.errors import CodexAppServerError
from codealmanac.integrations.harnesses.codex.events import (
    CodexRunState,
    done_event,
    map_codex_notification,
    provider_session_event,
)
from codealmanac.integrations.harnesses.codex.fields import (
    JsonObject,
    as_record,
    string_field,
)
from codealmanac.integrations.harnesses.codex.responses import (
    noninteractive_response,
)
from codealmanac.integrations.harnesses.codex.rpc import (
    JsonRpcLineProcess,
    is_server_request,
)
from codealmanac.integrations.harnesses.codex.run_result import (
    failed_result,
    result_from_state,
)
from codealmanac.integrations.harnesses.codex.sandbox import (
    SandboxMode,
    resolve_sandbox_mode,
    sandbox_policy,
)
from codealmanac.integrations.harnesses.codex.timeouts import env_milliseconds
from codealmanac.integrations.harnesses.codex.turn_completion import (
    root_turn_completion,
)
from codealmanac.services.harnesses.models import (
    HarnessEvent,
    HarnessRunResult,
)
from codealmanac.services.harnesses.requests import RunHarnessRequest

CODEX_APP_SERVER_RPC_TIMEOUT_MS = 30_000
CODEX_APP_SERVER_TURN_TIMEOUT_MS = 30 * 60_000
CODEX_APP_SERVER_RPC_TIMEOUT_ENV = "CODEALMANAC_CODEX_APP_SERVER_RPC_TIMEOUT_MS"
CODEX_APP_SERVER_TURN_TIMEOUT_ENV = "CODEALMANAC_CODEX_APP_SERVER_TURN_TIMEOUT_MS"


class CodexAppServerClient:
    def __init__(
        self,
        command: str = "codex",
        rpc_timeout_seconds: float | None = None,
        turn_timeout_seconds: float | None = None,
        sandbox_mode: SandboxMode | None = None,
    ):
        self.command = command
        self.rpc_timeout_seconds = rpc_timeout_seconds
        self.turn_timeout_seconds = turn_timeout_seconds
        self.sandbox_mode = sandbox_mode

    def run(self, request: RunHarnessRequest) -> HarnessRunResult:
        try:
            return self.run_once(request)
        except FileNotFoundError:
            return failed_result("codex not found on PATH")
        except CodexAppServerError as error:
            return failed_result(str(error))

    def run_once(self, request: RunHarnessRequest) -> HarnessRunResult:
        state = CodexRunState()
        events: list[HarnessEvent] = []
        rpc_timeout = self.resolved_rpc_timeout_seconds()
        turn_timeout = self.resolved_turn_timeout_seconds()
        sandbox_mode = self.resolved_sandbox_mode()
        process = JsonRpcLineProcess.start(
            self.command,
            ("app-server", "--config", "mcp_servers={}", "--listen", "stdio://"),
            request.cwd,
        )
        try:
            self.request_rpc(
                process,
                "initialize",
                {
                    "clientInfo": {
                        "name": "codealmanac",
                        "title": "CodeAlmanac",
                        "version": __version__,
                    },
                    "capabilities": {"experimentalApi": True},
                },
                rpc_timeout,
                state,
                events,
            )
            thread = as_record(
                self.request_rpc(
                    process,
                    "thread/start",
                    {
                        "cwd": str(request.cwd),
                        "model": request.model,
                        "approvalPolicy": "never",
                        "sandbox": sandbox_mode,
                        "developerInstructions": None,
                        "ephemeral": True,
                    },
                    rpc_timeout,
                    state,
                    events,
                )
            )
            thread_id = string_field(as_record(thread.get("thread")), "id")
            if thread_id is None:
                raise CodexAppServerError(
                    "Codex app-server thread/start did not return a thread id"
                )
            state.provider_session_id = thread_id
            state.root_thread_id = thread_id
            events.append(provider_session_event(thread_id))
            turn = as_record(
                self.request_rpc(
                    process,
                    "turn/start",
                    {
                        "threadId": thread_id,
                        "cwd": str(request.cwd),
                        "input": [
                            {
                                "type": "text",
                                "text": request.prompt,
                                "text_elements": [],
                            }
                        ],
                        "approvalPolicy": "never",
                        "sandboxPolicy": sandbox_policy(request.cwd, sandbox_mode),
                        "model": request.model,
                        "effort": None,
                        "outputSchema": None,
                    },
                    rpc_timeout,
                    state,
                    events,
                )
            )
            state.root_turn_id = string_field(as_record(turn.get("turn")), "id")
            return self.read_turn(process, turn_timeout, state, events)
        finally:
            process.terminate()

    def request_rpc(
        self,
        process: "JsonRpcLineProcess",
        method: str,
        params: JsonObject,
        timeout_seconds: float,
        state: CodexRunState,
        events: list[HarnessEvent],
    ) -> JsonValue:
        request_id = process.next_request_id()
        process.write({"id": request_id, "method": method, "params": params})
        deadline = time.monotonic() + timeout_seconds
        while True:
            message = process.read_until(deadline, f"{method} timed out")
            if is_server_request(message):
                self.respond_to_server_request(process, message)
                continue
            if "id" in message:
                if message.get("id") != request_id:
                    continue
                error = as_record(message.get("error"))
                if len(error) > 0:
                    detail = string_field(error, "message") or "request failed"
                    raise CodexAppServerError(f"Codex app-server {method}: {detail}")
                return message.get("result")
            if string_field(message, "method") is not None:
                events.extend(map_codex_notification(message, state))

    def read_turn(
        self,
        process: "JsonRpcLineProcess",
        timeout_seconds: float,
        state: CodexRunState,
        events: list[HarnessEvent],
    ) -> HarnessRunResult:
        deadline = time.monotonic() + timeout_seconds
        while True:
            message = process.read_until(deadline, "turn timed out")
            if is_server_request(message):
                self.respond_to_server_request(process, message)
                continue
            method = string_field(message, "method")
            if method is None:
                continue
            is_root_completion = root_turn_completion(message, state)
            mapped = map_codex_notification(message, state, is_root_completion)
            events.extend(mapped)
            if method == "error":
                return result_from_state(state, events)
            if method == "turn/completed" and is_root_completion:
                events.append(done_event(state))
                return result_from_state(state, events)

    def respond_to_server_request(
        self,
        process: "JsonRpcLineProcess",
        message: JsonObject,
    ) -> None:
        request_id = message.get("id")
        method = string_field(message, "method")
        if request_id is None or method is None:
            return
        response = noninteractive_response(method)
        if response.error is not None:
            process.write({"id": request_id, "error": response.error})
            return
        process.write({"id": request_id, "result": response.result})

    def resolved_rpc_timeout_seconds(self) -> float:
        if self.rpc_timeout_seconds is not None:
            return self.rpc_timeout_seconds
        return env_milliseconds(
            CODEX_APP_SERVER_RPC_TIMEOUT_ENV,
            CODEX_APP_SERVER_RPC_TIMEOUT_MS,
        ) / 1000

    def resolved_turn_timeout_seconds(self) -> float:
        if self.turn_timeout_seconds is not None:
            return self.turn_timeout_seconds
        return env_milliseconds(
            CODEX_APP_SERVER_TURN_TIMEOUT_ENV,
            CODEX_APP_SERVER_TURN_TIMEOUT_MS,
        ) / 1000

    def resolved_sandbox_mode(self) -> SandboxMode:
        return resolve_sandbox_mode(self.sandbox_mode)
